docs/getting-started/throwing-errors.md

```diff
@@ -0,0 +1,289 @@
+    * [Adding more data](#adding-more-data)
+    * [Being specific](#being-specific)
+
+
+## Types of error
+
+Errors can be sorted into two broad categories, it's important to understand the difference because they should be handled differently by your application. These take inspiration from the [Joyent error handling documentation](https://www.joyent.com/node-js/production/design/errors), which is also worth a read.
+
+### Operational errors
+
+An error can be considered "Operational" when they are _expected_ to occur as part of day-to-day running of an application. They normally represent that something went wrong outside of the control of the application code, e.g. a remote service failed. You can also consider an error operational when for things like end-user mistakes, e.g. filling out a form incorrectly.
+
+Some operational errors may be recoverable (as in the page can continue to render even if they occur). Some operational errors may not be recoverable, and in the context of a web service these will result in a non-`200` status code for an end-user. Both of these cases are still "Operational" if they're part of the expected running of the app.
+
+Operational errors should never result in an application crashing unexpectedly.
+
+### Programmer errors
+
+An error is considered to be a "Programmer Error" (or "Non-Operational") if it is unexpected and occurs because of a bug in the code. This category of error can always be avoided or fixed by making a code change. These errors can not be handled properly since the code in question is broken, e.g. `tried to call property X of "undefined"` – this is a result of code that doesn't properly validate the data it handles.
+
+
+## Types of error in code
+
+Reliability Kit provides a package to help us differentiate between Operational (known) errors, and Programmer (unknown) errors: [`@dotcom-reliability-kit/errors`](https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/errors#readme).
+
+The `OperationalError` class (explained fully in the [package documentation](https://github.com/Financial-Times/dotcom-reliability-kit/tree/main/packages/errors#operationalerror)) allows you to throw errors which are explicitly marked as "known". Any errors created with this constructor will include an `isOperational` property set to `true`, which can be used by our tooling later to help us understand the health of a system.
+
+```js
+throw new OperationalError({
+    code: 'EXAMPLE_ERROR',
+    message: 'This is an operational error'
+});
+```
+
+This (and further error classes that which extend it) should ideally be used _anywhere_ where you understand how to handle an error. This means in your app you might do something like this:
+
+```js
+// E.g. GET https://your-app/pokemon/oddish
+app.get('/pokemon/:name', async (request, response, next) => {
+    try {
+        // We make a request to an API, we understand that sometimes this API will fail
+        const response = await fetch(
```

Comment:
**suggestion:** as a good practice and from a standardisation POV, all the downstream API calls should be moved to a [helper function](https://github.com/Financial-Times/next-newsletters-page/blob/abd26fbb5233b75b1b4f56b76c5adf1f336d7666/server/lib/helpers.js#L3) so that the [error scenarios](https://github.com/Financial-Times/next-newsletters-page/blob/abd26fbb5233b75b1b4f56b76c5adf1f336d7666/server/lib/helpers.js#L29) and [validation errors](https://github.com/Financial-Times/next-newsletters-page/blob/abd26fbb5233b75b1b4f56b76c5adf1f336d7666/server/lib/helpers.js#L10) should be propagated back to the [calling function](https://github.com/Financial-Times/next-newsletters-page/blob/abd26fbb5233b75b1b4f56b76c5adf1f336d7666/server/controllers/displayPage.js#L14). the benefits of doing this will be abstracting the error handling scenarios to a common helper function and handling the errors back in the [catch section](https://github.com/Financial-Times/next-newsletters-page/blob/abd26fbb5233b75b1b4f56b76c5adf1f336d7666/server/controllers/displayPage.js#L97) (avoiding multiple try-catch block for multiple API calls).

Reply:
Oh also I couldn't help but notice in [this example](https://github.com/Financial-Times/next-newsletters-page/blob/abd26fbb5233b75b1b4f56b76c5adf1f336d7666/server/lib/helpers.js#L10) that your app is throwing a plain object with a `message` property rather than an `Error`. This means that any errors which get logged will not be able to include an error `stack` which is really helpful in debugging.

I'm going to add a section to the documentation which explains this, but in the meantime the two ESLint rules related to throwing and rejecting non-`Error` objects are useful:

  * https://eslint.org/docs/rules/no-throw-literal
  * https://eslint.org/docs/rules/prefer-promise-reject-errors

Reply:
> I don't want to dictate _everything_ that we do with error handling

I wouldn't say it is dictating, you can have it like: here are the two ways of doing it and chose your way. Obviously, there are tradeoffs between both approaches. so the consumer can decide on the approach

Reply:
I've now documented this in the [throwing errors](https://github.com/Financial-Times/dotcom-reliability-kit/blob/getting-started-guide/docs/getting-started/throwing-errors.md#using-error-objects) guide, thanks for the suggestion!